packages/dev/loaders/src/glTF/2.0/glTFLoader.ts

```diff
@@ -2774,25 +2774,44 @@ export class GLTFLoader implements IGLTFLoader {
         }
     }
 
-    private _applyExtensions<T>(property: IProperty, functionName: string, actionAsync: (extension: IGLTFLoaderExtension) => Nullable<T> | undefined): Nullable<T> {
-        for (const extension of this._extensions) {
-            if (extension.enabled) {
-                const id = `${extension.name}.${functionName}`;
-                const loaderProperty = property as ILoaderProperty;
-                loaderProperty._activeLoaderExtensionFunctions = loaderProperty._activeLoaderExtensionFunctions || {};
-                const activeLoaderExtensionFunctions = loaderProperty._activeLoaderExtensionFunctions;
-                if (!activeLoaderExtensionFunctions[id]) {
-                    activeLoaderExtensionFunctions[id] = true;
-
-                    try {
-                        const result = actionAsync(extension);
-                        if (result) {
-                            return result;
-                        }
-                    } finally {
-                        delete activeLoaderExtensionFunctions[id];
+    private _getActiveExtensionsHasOwnFunction<F extends IGLTFLoaderExtensionFunctionName>(functionName: F): IGLTFLoaderExtensionWithRequireFunction<F>[] {
+        let extensions = this._activeExtensionsHasOwnFunction[functionName];
+
+        if (!extensions) {
+            extensions = [];
+            for (const extension of this._extensions) {
+                if (extension[functionName] && extension.enabled) {
+                    extensions.push(extension as IGLTFLoaderExtensionWithRequireFunction<F>);
+                }
+            }
+        }
+
+        return extensions;
+    }
+
+    private _applyExtensions<T, F extends IGLTFLoaderExtensionFunctionName>(
+        property: IProperty,
+        functionName: F,
+        actionAsync: (extension: IGLTFLoaderExtensionWithRequireFunction<F>) => Nullable<T> | undefined
+    ): Nullable<T> {
+        const loaderProperty = property as ILoaderProperty;
+        loaderProperty._activeLoaderExtensionFunctions = loaderProperty._activeLoaderExtensionFunctions || new Map();
+        const activeLoaderExtensionFunctions = loaderProperty._activeLoaderExtensionFunctions;
+
+        if (!activeLoaderExtensionFunctions.has(functionName)) {

```

Comment:
Unfortunately, this won't work. Indexing by only `functionName` and not `${extension.name}.${functionName}` means that, if an object needs the same function applied to it from several extensions, only one will be applied.
E.g., a material using `KHR_materials_volume` will always also have either `KHR_materials_transmission` or `KHR_materials_diffuse_transmission`, both of which use the function `_loadMaterialPropertiesAsync`.
When fixing this issue, try testing with [this asset](https://github.com/KhronosGroup/glTF-Sample-Assets/blob/main/Models/CompareVolume/glTF-Binary/CompareVolume.glb). If it looks correct when loaded in the Sandbox, you've probably solved the issue.

I do agree that we should find a way to avoid recursively applying extensions, since it results in unexpected behavior with the order in which extensions are applied. E.g., we expect that KHR_materials_volume is applied after the two transmission extensions mentioned above. However, if you look at the `order` property of these 3 extensions, you'll see that it's wrong (it's reversed) and shouldn't possibly work.
But the recursion just conveniently masks this problem by un-reversing the order, thus making it correct. AKA it's two bugs cancelling each other out :)

Reply:
Oh, forgot to clarify: 
You will need to update the `order` of extensions once the recursion has been removed. 
At minimum, we know that volume will need to have an order higher than both transmission and diffuse transmission.